File: ui/base/src/main/java/org/cru/godtools/base/ui/activity/BaseActivity.java
package org.cru.godtools.base.ui.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;

import com.google.common.base.Objects;

import org.cru.godtools.base.ui.R2;
import org.greenrobot.eventbus.EventBus;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import butterknife.BindView;
import butterknife.ButterKnife;

import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP;
import static android.content.Intent.FLAG_ACTIVITY_SINGLE_TOP;

public abstract class BaseActivity extends AppCompatActivity {
    private static final String EXTRA_LAUNCHING_COMPONENT = "org.cru.godtools.BaseActivity.launchingComponent";

    // App/Action Bar
    @Nullable
    @BindView(R2.id.appbar)
    protected Toolbar mToolbar;
    @Nullable
    protected ActionBar mActionBar;

    @NonNull
    protected /*final*/ EventBus mEventBus;

    @NonNull
    public static Bundle buildExtras(@NonNull final Context context) {
        final Bundle extras = new Bundle();
        if (context instanceof Activity) {
            extras.putParcelable(EXTRA_LAUNCHING_COMPONENT, ((Activity) context).getComponentName());
        }
        return extras;
    }

    // region Lifecycle Events

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mEventBus = EventBus.getDefault();
    }

    @Override
    @CallSuper
    public void onContentChanged() {
        super.onContentChanged();

        // HACK: manually trigger this ButterKnife view binding to work around an inheritance across libraries bug
        // HACK: see: https://github.com/JakeWharton/butterknife/issues/808
        new BaseActivity_ViewBinding(this);

        ButterKnife.bind(this);
        setupActionBar();
    }

    @Override
    protected void onNewIntent(final Intent intent) {
        super.onNewIntent(intent);

        // update the Launching Component extra
        getIntent()
                .putExtra(EXTRA_LAUNCHING_COMPONENT, (Parcelable) intent.getParcelableExtra(EXTRA_LAUNCHING_COMPONENT));
    }

    @CallSuper
    protected void onSetupActionBar() {}

    // endregion Lifecycle Events

    private void setupActionBar() {
        if (mToolbar != null) {
            setSupportActionBar(mToolbar);
        }
        mActionBar = getSupportActionBar();
        if (mActionBar != null) {
            mActionBar.setDisplayHomeAsUpEnabled(true);
        }

        // trigger lifecycle event for subclasses
        onSetupActionBar();
    }

    // region Up Navigation

    @Override
    public void supportNavigateUpTo(@NonNull final Intent upIntent) {
        // if the upIntent already points to the original launching activity, just finish this activity
        if (Objects.equal(getIntent().getParcelableExtra(EXTRA_LAUNCHING_COMPONENT), upIntent.getComponent())) {
            finish();
            return;
        }

        // otherwise defer to default navigate behavior
        super.supportNavigateUpTo(upIntent);
    }

    @Nullable
    @Override
    public Intent getSupportParentActivityIntent() {
        final Intent intent = super.getSupportParentActivityIntent();

        if (intent != null) {
            intent.addFlags(FLAG_ACTIVITY_SINGLE_TOP | FLAG_ACTIVITY_CLEAR_TOP);
            intent.putExtras(buildParentIntentExtras());
        }

        return intent;
    }

    @NonNull
    @CallSuper
    protected Bundle buildParentIntentExtras() {
        return new Bundle();
    }

    // endregion Up Navigation
}

File: ui/base/src/main/kotlin/org/cru/godtools/base/ui/activity/BaseActivity.kt
package org.cru.godtools.base.ui.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.os.Parcelable
import androidx.annotation.CallSuper
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import javax.inject.Inject
import org.ccci.gto.android.common.androidx.lifecycle.onDestroy
import org.ccci.gto.android.common.androidx.lifecycle.onResume
import org.ccci.gto.android.common.compat.content.getParcelableExtraCompat
import org.cru.godtools.base.Settings
import org.greenrobot.eventbus.EventBus

private const val EXTRA_FEATURE_DISCOVERY = "org.cru.godtools.BaseActivity.FEATURE_DISCOVERY"
private const val EXTRA_FEATURE = "org.cru.godtools.BaseActivity.FEATURE"
private const val EXTRA_FORCE = "org.cru.godtools.BaseActivity.FORCE"
private const val EXTRA_LAUNCHING_COMPONENT = "org.cru.godtools.BaseActivity.launchingComponent"

@VisibleForTesting
internal const val MSG_FEATURE_DISCOVERY = 1

abstract class BaseActivity protected constructor() : AppCompatActivity() {
    @Inject
    protected lateinit var eventBus: EventBus
    @Inject
    protected lateinit var settings: Settings

    // region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupFeatureDiscovery(savedInstanceState)
    }

    override fun onNewIntent(newIntent: Intent) {
        super.onNewIntent(newIntent)

        // update the Launching Component extra
        intent.putExtra(
            EXTRA_LAUNCHING_COMPONENT,
            newIntent.getParcelableExtraCompat(EXTRA_LAUNCHING_COMPONENT, Parcelable::class.java)
        )
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.saveFeatureDiscoveryState()
    }
    // endregion Lifecycle

    // region Feature Discovery
    protected var featureDiscoveryActive: String? = null

    private fun setupFeatureDiscovery(savedInstanceState: Bundle?) {
        savedInstanceState?.restoreFeatureDiscoveryState()
        lifecycle.onResume { triggerFeatureDiscovery() }
    }

    private fun triggerFeatureDiscovery() {
        when (val feature = featureDiscoveryActive) {
            null -> showNextFeatureDiscovery()
            else -> showFeatureDiscovery(feature, true)
        }
    }

    protected open fun showNextFeatureDiscovery() = Unit

    protected fun showFeatureDiscovery(feature: String, force: Boolean = false) {
        // short-circuit if this activity is not started
        if (!lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) return

        // short-circuit if feature discovery is already visible
        if (isFeatureDiscoveryVisible()) return

        // short-circuit if this feature was discovered and we aren't forcing it
        if (settings.isFeatureDiscovered(feature) && !force) return

        // short-circuit if we can't show this feature discovery right now,
        // and try to show the next feature discovery that can be shown.
        if (!canShowFeatureDiscovery(feature)) {
            showNextFeatureDiscovery()
            return
        }

        // actually show the feature
        onShowFeatureDiscovery(feature, force)
    }

    @CallSuper
    protected open fun onShowFeatureDiscovery(feature: String, force: Boolean) = Unit

    /**
     * @return true if the activity is in a state that it can actually show the specified feature discovery.
     */
    @CallSuper
    protected open fun canShowFeatureDiscovery(feature: String) = true

    @CallSuper
    protected open fun isFeatureDiscoveryVisible() = false

    private fun Bundle.saveFeatureDiscoveryState() {
        putString(EXTRA_FEATURE_DISCOVERY, featureDiscoveryActive)
    }

    private fun Bundle.restoreFeatureDiscoveryState() {
        featureDiscoveryActive = getString(EXTRA_FEATURE_DISCOVERY, featureDiscoveryActive)
    }

    // region Delayed Dispatch
    @VisibleForTesting
    internal val featureDiscoveryHandler by lazy {
        Handler(mainLooper) { m -> showFeatureDiscovery(m) }
            .apply { lifecycle.onDestroy { removeCallbacksAndMessages(null) } }
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
    fun dispatchDelayedFeatureDiscovery(feature: String, force: Boolean, delay: Long) {
        val message = featureDiscoveryHandler.obtainMessage(MSG_FEATURE_DISCOVERY, feature).apply {
            data = Bundle().apply {
                putString(EXTRA_FEATURE, feature)
                putBoolean(EXTRA_FORCE, force)
            }
        }
        featureDiscoveryHandler.sendMessageDelayed(message, delay)
    }

    private fun showFeatureDiscovery(message: Message): Boolean {
        message.data.getString(EXTRA_FEATURE)?.let {
            showFeatureDiscovery(it, message.data.getBoolean(EXTRA_FORCE, false))
        }
        return true
    }

    protected fun purgeQueuedFeatureDiscovery(feature: String) {
        featureDiscoveryHandler.removeMessages(MSG_FEATURE_DISCOVERY, feature)
    }
    // endregion Delayed Dispatch
    // endregion Feature Discovery

    // region Up Navigation
    override fun supportNavigateUpTo(upIntent: Intent) {
        // if the upIntent already points to the original launching activity, just finish this activity
        if (upIntent.component == intent.getParcelableExtraCompat(EXTRA_LAUNCHING_COMPONENT, Parcelable::class.java)) {
            finish()
            return
        }

        // otherwise defer to default navigate behavior
        super.supportNavigateUpTo(upIntent)
    }

    override fun getSupportParentActivityIntent() = super.getSupportParentActivityIntent()?.apply {
        addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        putExtras(buildParentIntentExtras())
    }

    @CallSuper
    protected open fun buildParentIntentExtras() = Bundle()
    // endregion Up Navigation

    companion object {
        fun buildExtras(context: Context) = Bundle().apply {
            if (context is Activity) putParcelable(EXTRA_LAUNCHING_COMPONENT, context.componentName)
        }
    }
}
